win32ss/user/ntuser/msgqueue.c

```diff
@@ -1870,46 +1871,74 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
+
+                     change_language = TRUE;
+                  }
+                  else if ( Msg->wParam == VK_CONTROL && Msg->wParam == language_hotkey
+                           && !(IS_KEY_DOWN(gafAsyncKeyState, VK_CONTROL)))
+                  {
+                     if (IS_KEY_DOWN(gafAsyncKeyState, VK_LSHIFT))
+                        gLanguageToggleKeyState = INPUTLANGCHANGE_FORWARD;  // Left Alt - Left Shift, Next
+                     else if (IS_KEY_DOWN(gafAsyncKeyState, VK_RSHIFT))
+                        gLanguageToggleKeyState = INPUTLANGCHANGE_BACKWARD; // Left Alt - Right Shift, Previous
+
+                     change_language = TRUE;
+                  }
+               }
+            }
+        }
 
-        if ( gdwLanguageToggleKey < 3 &&
-             IS_KEY_DOWN(gafAsyncKeyState, gdwLanguageToggleKey == 1 ? VK_LMENU : VK_CONTROL) )
+        if ( change_language )
         {
-            if ( Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT)))
+            WPARAM wParamILR;
+            PKL pkl = pti->KeyboardLayout;
+
+            if (pWnd) UserDerefObjectCo(pWnd);
+
+            //// Seems to override message window.
+            if (!(pWnd = pti->MessageQueue->spwndFocus))
             {
-                WPARAM wParamILR;
-                PKL pkl = pti->KeyboardLayout;
+               pWnd = pti->MessageQueue->spwndActive;
+            }
+            if (pWnd) UserRefObjectCo(pWnd, &Ref);
 
-                if (pWnd) UserDerefObjectCo(pWnd);
+            if (pkl != NULL && gLanguageToggleKeyState)
+            {
+               TRACE("Posting WM_INPUTLANGCHANGEREQUEST KeyState %d\n", gLanguageToggleKeyState );
 
-                //// Seems to override message window.
-                if (!(pWnd = pti->MessageQueue->spwndFocus))
-                {
-                     pWnd = pti->MessageQueue->spwndActive;
-                }
-                if (pWnd) UserRefObjectCo(pWnd, &Ref);
+               wParamILR = gLanguageToggleKeyState;
+               // If system character set and font signature send flag.
+               if ( gSystemFS & pkl->dwFontSigs )
+               {
+                  wParamILR |= INPUTLANGCHANGE_SYSCHARSET;
+               }
 
-                if (pkl != NULL && gLanguageToggleKeyState)
-                {
-                    TRACE("Posting WM_INPUTLANGCHANGEREQUEST KeyState %d\n", gLanguageToggleKeyState );
-
-                    wParamILR = gLanguageToggleKeyState;
-                    // If system character set and font signature send flag.
-                    if ( gSystemFS & pkl->dwFontSigs )
-                    {
-                       wParamILR |= INPUTLANGCHANGE_SYSCHARSET;
-                    }
-
-                    UserPostMessage( UserHMGetHandle(pWnd),
-                                     WM_INPUTLANGCHANGEREQUEST,
-                                     wParamILR,
-                                    (LPARAM)pkl->hkl );
-
```

Comment:
Please reintroduce the new line after this `UserPostMessage( WM_INPUTLANGCHANGEREQUEST )` call.

Reply:
Adjusted other places too